In [1]:
import pandas as pd

In [2]:
import csv

### Allegheny County Employee Salaries firstly we have to find out how many employees are working each department and this can easily be extracted through pandas through value counts methods

In [3]:
salary_pd=pd.read_csv("redacted-2022-december-31-wprdc.csv")

In [4]:
salary_pd.head()

FIRST_NAME LAST_NAME            Combo Name             DEPARTMENT  \
0    CATHERINE     ABALO      ABALO, CATHERINE  Kane Regional Centers   
1      KEVIN J    ABBOTT       ABBOTT, KEVIN J   Emergency Management   
2        JOY M    ABBOTT         ABBOTT, JOY M  Kane Regional Centers   
3  ELIZABETH S   ABRAHAM  ABRAHAM, ELIZABETH S                  Parks   
4      JASMINE     ABRAM        ABRAM, JASMINE   Emergency Management   

                     JOB_TITLE  ELECTED_OFFICIAL DATE_STARTED SEX  \
0            NURSING ASSISTANT                 0     4/8/2010   F   
1  FIRE INSTRUCTOR - PART TIME                 0    6/11/2018   M   
2                         COOK                 0    2/14/1999   F   
3                SEASONAL AIDE                 0    5/26/2022   F   
4    TELECOMMUNICATION OFFICER                 0    5/23/2022   F   

                        ETHNICITY ORIG_START DATE_TERM  PAY_STATUS  \
0                           Black   7/6/2009       NaN      Active   
1  White (Not of Hispanic Origin)  6/11/2018       NaN      Active   
2  White (Not of Hispanic Origin)   2/2/1998       NaN      Active   
3  White (Not of Hispanic Origin)  5/26/2022  9/5/2022  Terminated   
4                           Black  5/31/2016       NaN      Active   

   ANNUAL_SALARY   REGULAR_PAY   OVERTIME_PAY   INCENTIVE_PAY    GROSS_PAY   \
0      41,581.07     40,740.50      23,416.98          125.00    64,282.48    
1      30,576.00        378.00            -               -         378.00    
2      40,761.76     44,015.85       1,923.77          225.00    46,164.62    
3      39,936.00      6,488.00            -            100.00     6,588.00    
4      52,240.03     25,637.96      15,994.95          125.00    41,757.91    

  Unnamed: 17  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [5]:
salary_pd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6280 entries, 0 to 6279
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   FIRST_NAME        6280 non-null   object
 1   LAST_NAME         6280 non-null   object
 2   Combo Name        6280 non-null   object
 3   DEPARTMENT        6280 non-null   object
 4   JOB_TITLE         6280 non-null   object
 5   ELECTED_OFFICIAL  6280 non-null   int64 
 6   DATE_STARTED      6280 non-null   object
 7   SEX               6280 non-null   object
 8   ETHNICITY         6280 non-null   object
 9   ORIG_START        6280 non-null   object
 10  DATE_TERM         1263 non-null   object
 11  PAY_STATUS        6280 non-null   object
 12   ANNUAL_SALARY    6280 non-null   object
 13   REGULAR_PAY      6280 non-null   object
 14   OVERTIME_PAY     6280 non-null   object
 15   INCENTIVE_PAY    6280 non-null   object
 16   GROSS_PAY        6280 non-null   object
 17  Unnamed: 17   

In [6]:
with open("redacted-2022-december-31-wprdc.csv") as f:
    salary=list(csv.reader(f))
    header=salary[0]
    salary=salary[1:]

In [7]:
salary_pd["DEPARTMENT"].value_counts()

DEPARTMENT
Human Services             935
Kane Regional Centers      814
Parks                      732
Jail                       635
Emergency Management       404
Health                     401
Police                     298
District Attorney          266
Public Works Operating     240
Facilities Management      230
Administrative Services    217
Sheriff                    210
Public Defender            162
Court Records              122
Information Technology     115
Medical Examiner           105
Controller                  92
Treasurer                   79
Economic Development        78
County Solicitor            57
Human Resources             27
County Manager              20
Budget & Finance             9
Equity and Inclusion         9
County Council               7
Children Initiatives         7
Retirement System            4
Sustainability               3
County Executive             2
Name: count, dtype: int64

In [8]:
##Now we have to identify the how many job titile are under our data set.

In [9]:
salary_pd["JOB_TITLE"].value_counts().head(50)

JOB_TITLE
SEASONAL AIDE                     552
CORRECTIONAL OFFICER              406
TELECOMMUNICATION OFFICER         314
LABORER                           195
POLICE OFFICER                    173
NURSING ASSISTANT                 153
DEPUTY SHERIFF                    149
CASEWORKER SUPERVISOR              95
CASEWORKER - CYF (lng Step)        79
SERGEANT                           76
STUDENT INTERN                     67
LICENSED PRACTICAL NURSE           64
BUILDING GUARD                     62
CASEWORKER 5 - 10 - CYF            60
ATTORNEY 4                         59
CASEWORKER S - 3 - CYF             59
ADMINISTRATIVE ASSISTANT           56
CASEWORKER - CYF                   55
DRIVER 002 WITH CDL                53
RN RESIDENT CARE COORDINATOR       53
ASSISTANT COUNTY SOLICITOR         47
CUSTODIAL WORKER                   44
CLERK 5                            42
ATTORNEY 3                         42
SCIENTIST                          40
CLERK TYPIST 12 YEAR               37
HO

In [10]:
salary_pd["PAY_STATUS"].value_counts()

PAY_STATUS
Active        5017
Terminated    1263
Name: count, dtype: int64

## so we have 5017 employee active so will work on active employee now we require how many employee are male and female in active status and who is highest payer gross salary and highest payer in male and female and what is the highest paying department.

In [11]:
salary_pd["PAY_STATUS"]=="Active"


0        True
1        True
2        True
3       False
4        True
        ...  
6275     True
6276     True
6277     True
6278     True
6279     True
Name: PAY_STATUS, Length: 6280, dtype: bool

In [12]:
Active=salary_pd.loc[salary_pd["PAY_STATUS"]=="Active"]


In [13]:
Active["SEX"].value_counts()

SEX
M    2620
F    2397
Name: count, dtype: int64

In [14]:
salary_pd["ANNUAL_SALARY"]

KeyError: 'ANNUAL_SALARY'

In [15]:
print(salary_pd.columns)

Index(['FIRST_NAME', 'LAST_NAME', 'Combo Name', 'DEPARTMENT', 'JOB_TITLE',
       'ELECTED_OFFICIAL', 'DATE_STARTED', 'SEX', 'ETHNICITY', 'ORIG_START',
       'DATE_TERM', 'PAY_STATUS', ' ANNUAL_SALARY ', ' REGULAR_PAY ',
       ' OVERTIME_PAY ', ' INCENTIVE_PAY ', ' GROSS_PAY ', 'Unnamed: 17'],
      dtype='object')


In [16]:
salary_pd.columns=salary_pd.columns.str.strip()

In [17]:
Active.columns=Active.columns.str.strip()

In [18]:
Active["GROSS_PAY"]=Active["GROSS_PAY"].str.replace(',', '').astype(float)

C:\Users\user\AppData\Local\Temp\ipykernel_6356\1632086882.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Active["GROSS_PAY"]=Active["GROSS_PAY"].str.replace(',', '').astype(float)


In [19]:
Active.loc[:,["Combo Name","GROSS_PAY","SEX"]].min()   ## minimum salary 

Combo Name    ABALO, CATHERINE
GROSS_PAY                 63.0
SEX                          F
dtype: object

In [20]:
Active.loc[:,["Combo Name","GROSS_PAY","SEX"]].max()     ## Maximum Salary

Combo Name    ZWOLINSKI, STEVEN
GROSS_PAY             269356.28
SEX                           M
dtype: object

In [21]:
Active.loc[:,["Combo Name","GROSS_PAY","SEX"]]["GROSS_PAY"].mean()  ### Average Salary

59178.42286426151

In [22]:
Active["GROSS_PAY"].sum()

296898147.51

In [23]:
Active.pivot_table(values="GROSS_PAY", index="SEX", aggfunc="mean")

GROSS_PAY
SEX              
F    52469.153813
M    65316.635809

In [24]:
Active.pivot_table(values="GROSS_PAY", index="DEPARTMENT", aggfunc="mean") ## we can extract max and min salary 

GROSS_PAY
DEPARTMENT                            
Administrative Services   45356.198531
Budget & Finance          78462.197778
Children Initiatives      60941.928571
Controller                57922.996988
County Council            70300.792500
County Executive         162126.450000
County Manager            84574.780588
County Solicitor          60337.995686
Court Records             45331.767549
District Attorney         59858.273667
Economic Development      50563.250909
Emergency Management      61007.012102
Equity and Inclusion      51564.548571
Facilities Management     52378.975572
Health                    50121.654704
Human Resources           54789.137500
Human Services            57747.598257
Information Technology    63094.134712
Jail                      78324.737065
Kane Regional Centers     51287.185000
Medical Examiner          68137.151087
Parks                     24330.253050
Police                   101296.706250
Public Defender           49761.305071
Public Works Operating    55024.158380
Retirement System         63414.320000
Sheriff                   91444.324924
Sustainability            41460.023333
Treasurer                 49303.322121

In [25]:
Active.pivot_table(values="GROSS_PAY", index="ETHNICITY", aggfunc="mean")

GROSS_PAY
ETHNICITY                                   
American Indian/Alaskan Native  73348.943333
Asian or Pacific Islander       54114.603051
Black                           56038.215473
Hispanic                        61623.942564
Two or More Races               52142.180800
White (Not of Hispanic Origin)  59868.264019

In [26]:
Active.pivot_table(values="GROSS_PAY", index="JOB_TITLE", aggfunc="mean").max()

GROSS_PAY    236752.69
dtype: float64

#### it is observed that male are being paid high salary with average 65316 compared to female 52469 and similarly average highest salary is 162126.450000 in county executives department and highest paying ETHNICITY is American/Alaskan Native. 
 

In [27]:
import datetime as dt

In [28]:
Active["DATE_STARTED"]=pd.to_datetime(Active["DATE_STARTED"])

C:\Users\user\AppData\Local\Temp\ipykernel_6356\500098170.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Active["DATE_STARTED"]=pd.to_datetime(Active["DATE_STARTED"])


In [29]:
Active["ORIG_START"]=pd.to_datetime(Active["ORIG_START"])

C:\Users\user\AppData\Local\Temp\ipykernel_6356\1728814147.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Active["ORIG_START"]=pd.to_datetime(Active["ORIG_START"])


In [30]:
Active["DATE_STARTED"]-Active["ORIG_START"]

0       276 days
1         0 days
2       377 days
4      2183 days
5       237 days
          ...   
6275      0 days
6276      0 days
6277      0 days
6278    574 days
6279      0 days
Length: 5017, dtype: timedelta64[ns]

In [31]:
((Active["DATE_STARTED"]-Active["ORIG_START"]) != pd.Timedelta(0)).sum() 

1170

## so there are 1170 actve employees where there is difference between Date Started and origin start so need to be investigated further.

In [ ]:
# i need to extract the service period of active employees in terms of years and months and days closing date is 12/31/2023

In [40]:
pd.to_datetime('12/31/23', format='%m/%d/%y')

Timestamp('2023-12-31 00:00:00')

In [64]:

(pd.to_datetime('12/31/23', format='%m/%d/%y')) - Active["DATE_STARTED"]


0       5015 days
1       2029 days
2       9086 days
4        587 days
5       8400 days
          ...    
6275    9218 days
6276    6250 days
6277    3877 days
6278   14111 days
6279     412 days
Name: DATE_STARTED, Length: 5017, dtype: timedelta64[ns]